apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,311 @@
+            {
+                LogSessionsInUse();
+                SignalAnyWaitingRequests();
+                SessionPoolImpl targetPool = GlobalPool.GetOrAdd(result,
+                                             key => new SessionPoolImpl(key));
+
+                SpannerClient evictionClient = result.Client;
+                Session evictionSession = null;
+
+                //Figure out if we want to pool this released session.
+                lock (s_priorityListSync)
+                {
+                    targetPool.ReleaseSessionToPool(client, session);
+                    if (CurrentPooledSessions > MaximumPooledSessions)
+                    {
+                        var evictionPool = PriorityList.First().Value;
+                        evictionClient = evictionPool.Key.Client;
+                        evictionSession = evictionPool.AcquireEvictionCandidate();
+                        ReSortMru(evictionPool);
+                    }
+                    ReSortMru(targetPool);
+                }
+                if (evictionSession != null)
+                {
+                    Task.Run(() => EvictSession(evictionClient, evictionSession));
+                }
+            }
+        }
+
+        private static async Task EvictSession(SpannerClient evictionClient, Session evictionSession)
+        {
+            await evictionSession.RemoveFromTransactionPool().ConfigureAwait(false);
+            await evictionClient.DeleteSessionAsync(evictionSession.SessionName);
+        }
+
+        /// <summary>
+        /// Called to close a session and release it without putting it back into the pool.
+        /// </summary>
+        /// <param name="session"></param>
+        /// <returns></returns>
+        public static async Task Close(this Session session)
+        {
+            session.ThrowIfNull(nameof(session));
+
+            SessionPoolKey result;
+            if (s_sessionsInUse.TryRemove(session, out result))
+            {
+                LogSessionsInUse();
+                SignalAnyWaitingRequests();
+                await result.Client.DeleteSessionAsync(session.SessionName).ConfigureAwait(false);
+                return;
+            }
+            throw new InvalidOperationException("Close Session was not able to locate the provided session.");
+        }
+
+        private static void ReSortMru(SessionPoolImpl targetPool)
+        {
+            lock (s_priorityListSync)
+            {
+                if (PriorityList.Count > 1)
+                {
+                    PriorityList.Remove(targetPool);
+                    PriorityList.Add(targetPool, targetPool);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Releases all pooled sessions and frees resources on the server.
+        /// </summary>
+        /// <returns></returns>
+        public static Task ReleaseAll()
+        {
+            var entries = GlobalPool.Values;
```

Comment:
I would materialize this - it's not clear what happens to the previously-returned `Values` if you call `Clear` afterwards...